### PR TITLE
Update the encrypt and decrypt functions to support all PHP versions

### PR DESCRIPTION
**Pull request by hibatallahAouadni**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.6.1.x
| Description?  | add a test to check what kind of PHP version and function exists to choose the right function to encrypt or decrypt.
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/PSCSX-9164
| How to test?  | Please indicate how to best verify that this PR is correct.

### CODE REVIEW

File: classes/Rijndael.php [hunk 1]
```diff
@@ -48,24 +48,41 @@ public function encrypt($plaintext)
         if ($length >= 1048576) {
             return false;
         }
-        return base64_encode(mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $this->_key, $plaintext, MCRYPT_MODE_CBC, $this->_iv)).sprintf('%06d', $length);
+        $ciphertext = null;
+        if (function_exists('mcrypt_encrypt') && version_compare(phpversion(), '7.1', '<')) {
+            $ciphertext = mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $this->_key, $plaintext, MCRYPT_MODE_CBC, $this->_iv);
+        } elseif (function_exists('openssl_encrypt') && version_compare(phpversion(), '7.1', '>=')) {
```
Comment (jocel1) [inline classes/Rijndael.php]: why not testing first openssl_encrypt? you'd not need the version_compare part anymore
e.g:
if (function_exists('openssl_encrypt')) {

} elseif (function_exists('mcrypt_encrypt')) {

} else {
  throw new runtimeException(...);
}

File: classes/Rijndael.php [hunk 2]
```diff
@@ -48,24 +48,41 @@ public function encrypt($plaintext)
...
 
     public function decrypt($ciphertext)
     {
+        $output = null;
         if (ini_get('mbstring.func_overload') & 2) {
             $length = intval(mb_substr($ciphertext, -6, 6, ini_get('default_charset')));
             $ciphertext = mb_substr($ciphertext, 0, -6, ini_get('default_charset'));
+            if (function_exists('mcrypt_decrypt') && version_compare(phpversion(), '7.1', '<')) {
+                $output = mcrypt_decrypt(MCRYPT_RIJNDAEL_128, $this->_key, base64_decode($ciphertext), MCRYPT_MODE_CBC, $this->_iv);
+            } elseif (function_exists('openssl_decrypt') && version_compare(phpversion(), '7.1', '>=')) {
```
Comment (jocel1) [inline classes/Rijndael.php]: same here

File: classes/Rijndael.php [hunk 3]
```diff
@@ -48,24 +48,41 @@ public function encrypt($plaintext)
...
                 $length,
                 ini_get('default_charset')
             );
         } else {
             $length = intval(substr($ciphertext, -6));
             $ciphertext = substr($ciphertext, 0, -6);
-            return substr(mcrypt_decrypt(MCRYPT_RIJNDAEL_128, $this->_key, base64_decode($ciphertext), MCRYPT_MODE_CBC, $this->_iv), 0, $length);
+            if (function_exists('mcrypt_decrypt') && version_compare(phpversion(), '7.1', '<')) {
+                $output = mcrypt_decrypt(MCRYPT_RIJNDAEL_128, $this->_key, base64_decode($ciphertext), MCRYPT_MODE_CBC, $this->_iv);
+            } elseif (function_exists('openssl_decrypt') && version_compare(phpversion(), '7.1', '>=')) {
```
Comment (jocel1) [inline classes/Rijndael.php]: same here

**Review by jocel1 — CHANGES_REQUESTED** [review]

File: classes/Rijndael.php (hunk 1 above)
Comment (hibatallahAouadni) [reply classes/Rijndael.php]: you got a point :+1:
I'll do it, thanks :)

Comment (eternoendless) [reply classes/Rijndael.php]: Why test for PHP 7.1 ? From the looks of it you should only check for PHP 5.3.3 in order to make sure that openssl_encrypt has the $iv parameter

> 5.3.3 | The iv parameter was added.

**Review by eternoendless — CHANGES_REQUESTED** [review]

**Comment by hibatallahAouadni** [comment]
@jocel1 @eternoendless done :+1:

File: classes/Rijndael.php [hunk 4]
```diff
@@ -48,24 +48,47 @@ public function encrypt($plaintext)
         if ($length >= 1048576) {
             return false;
         }
-        return base64_encode(mcrypt_encrypt(MCRYPT_RIJNDAEL_128, $this->_key, $plaintext, MCRYPT_MODE_CBC, $this->_iv)).sprintf('%06d', $length);
+        $ciphertext = null;
+        if (function_exists('openssl_encrypt') && version_compare(phpversion(), '5.3.3', '>=')) {
+            $ciphertext = openssl_encrypt($plaintext, 'AES-128-CBC', $this->_key, OPENSSL_RAW_DATA, $this->_iv);
+        } elseif (function_exists('mcrypt_encrypt') && version_compare(phpversion(), '5.3.3', '<')) {
```
Comment (jocel1) [inline classes/Rijndael.php]: version_compare is not needed here

Comment (hibatallahAouadni) [reply classes/Rijndael.php]: @jocel1 you're right I'll change it :+1: 
sorry for the late response :/ I didn't get the notification.

**Comment by hibatallahAouadni** [comment]
@jocel1 @eternoendless thanks guys :)

**Review by eternoendless — APPROVED** [review]

**Review by jocel1 — APPROVED** [review]

**Comment by maximebiloe** [comment]
Thank you @hibatallahAouadni

**Comment by hakeryk2** [comment]
@eternoendless or @jocel1  Is this will be implemented in 1.6.1.23? It was merged for 1.6.1.17 but it didn't made up to the final  version and it's realy important for 7.2 and security.

**Comment by eternoendless** [comment]
@hakeryk2 This has changed since 1.6.1.20, see: https://github.com/PrestaShop/PrestaShop/pull/9218

**Comment by NoZTurn** [comment]
when using php7.3 with prestashop 1623,we can install normally,but cann't visit index.php and admin index, the error message is: Fatal error: Uncaught RuntimeException: Either Mcrypt or OpenSSL extension is required to run Prestashop in D:\wamp\www\prestashop1623\classes\Rijndael.php:68

**Pull request merged**